### PR TITLE
DM-46239:  Add pixelization envelopes for Intersection/UnionRegion

### PR DESCRIPTION
**Pull request by mwittgen**
## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: include/lsst/sphgeom/CompoundRegion.h [hunk 1]
```diff
@@ -88,6 +88,7 @@ class CompoundRegion : public Region {
     }
     static std::unique_ptr<CompoundRegion> decode(std::uint8_t const *buffer, size_t n);
     ///@}
+    static void flatten(Region const &region, std::vector<std::unique_ptr<Region>> &result);
```
Comment (erykoff) [inline include/lsst/sphgeom/CompoundRegion.h:91]: I don't think this should be here?

**Review by erykoff — COMMENTED** [review]
First: a typo and a leftover flatten.  Second: I'm worried that findPixels is recursive, I thought that there was discussion on the ticket that this is bad?  (This may be way out of my wheelhouse though).

File: include/lsst/sphgeom/CompoundRegion.h (hunk 1 above)
Comment (mwittgen) [reply include/lsst/sphgeom/CompoundRegion.h:91]: nope

**Comment by mwittgen** [comment]
Allowing to loop over a flattened structures, we wanted to push to another ticket when it's clearer how to define a better API.

File: src/PixelFinder.h [hunk 2]
```diff
@@ -156,23 +157,34 @@ template <
...
-            s, dynamic_cast<ConvexPolygon const &>(r), level, maxRanges);
+                s, *polygon, level, maxRanges);
         find();
+    } else if (auto union_region = dynamic_cast<UnionRegion const *>(&r)) {
+        Region const &region1 = union_region->getOperand(0);
+        Region const &region2 = union_region->getOperand(1);
+        auto rs1 = findPixels<Finder, InteriorOnly>(region1, maxRanges, level);
+        auto rs2 = findPixels<Finder, InteriorOnly>(region2, maxRanges, level);
+        s = rs1.join(rs2);
+    }else if (auto intersection_region = dynamic_cast<IntersectionRegion const *>(&r)) {
```
Comment (TallJimbo) [inline src/PixelFinder.h:180]:
```suggestion
    } else if (auto intersection_region = dynamic_cast<IntersectionRegion const *>(&r)) {
```

**Review by TallJimbo — APPROVED** [review]
Looks good.  I wish we had an alternative to a `getRegions` that deep-copies everything just to let Python iterate over it, but given pybind11's inflexibility re `unique_ptr` vs. `shared_ptr` holders for the same type, I imagine we're pretty much out of options.

**Comment by TallJimbo** [comment]
I think we're fine with letting this continue to be a fully recursive implementation and punt flattening that out to a new ticket.

**Pull request merged**